pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go

```diff
@@ -617,11 +637,45 @@ func (c *clusterAutoscaler) computeShootResourcesData(serviceAccountName string)
 			},
 		}
 	)
+	objects := []client.Object{clusterRole, clusterRoleBinding, role, rolebinding}
 
-	return registry.AddAllAndSerialize(
-		clusterRole,
-		clusterRoleBinding,
-		role,
-		rolebinding,
-	)
+	if workersHavePrioritiesConfigured {
+		configMap, err := c.generatePriorityExpanderConfigMap()
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, configMap)
+	}
+	return registry.AddAllAndSerialize(objects...)
+}
+
+func (c *clusterAutoscaler) generatePriorityExpanderConfigMap() (*corev1.ConfigMap, error) {
+	priorities := map[int32][]string{}
+	for _, machineDeployment := range c.machineDeployments {
+		priorities[machineDeployment.Priority] = append(priorities[machineDeployment.Priority], fmt.Sprintf("%s.%s", c.namespace, machineDeployment.Name))
+	}
+
+	priorityConfig, err := yaml.Marshal(priorities)
+	if err != nil {
+		return nil, err
+	}
+
+	configMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cluster-autoscaler-priority-expander",
+			Namespace: metav1.NamespaceSystem,
+		},
+		Data: map[string]string{
+			"priorities": string(priorityConfig),
+		},
+	}
+
+	return configMap, nil
+}
+
+func ensureExpanderInExpanderConfig(expander string, expanderConfig gardencorev1beta1.ExpanderMode) gardencorev1beta1.ExpanderMode {
+	if strings.Contains(expander, string(expanderConfig)) {
```

Comment:
This should be the other way around :)
```suggestion
	if strings.Contains(string(expanderConfig), expander) {
```

Could we also add a small test case to cover this function.

Reply:
Sorry for that!

I see two options for creating this test case:
1. Export the function and explicitly test it
2. Introduce even more variables in the existing `Deploy` test.

Against the `1.` stands the argument that, other than testing, we do not _need_ to export the function.
But on the other hand `2.` is not a very beautiful solution as well, as the test is already quite complicated and this would make it even harder to understand IMO.

Am I missing a third option / what do you think?

Reply:
I was thinking something more towards `2.`. You're right that the test has some complexity issues, but a possible change like this doesn't seem too complex, I think:
```diff
diff --git a/pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go b/pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go
index 52f0bb3fd6..19b4525e41 100644
--- a/pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go
+++ b/pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go
@@ -257,7 +257,7 @@ var _ = Describe("ClusterAutoscaler", func() {
                        },
                        Type: corev1.SecretTypeOpaque,
                }
-               deploymentFor = func(withConfig bool, withWorkerPriority bool) *appsv1.Deployment {
+               deploymentFor = func(withConfig, withWorkerPriority bool) *appsv1.Deployment {
                        var commandConfigFlags []string
 
                        expander := string(configExpander)
@@ -633,11 +633,14 @@ var _ = Describe("ClusterAutoscaler", func() {
 
        Describe("#Deploy", func() {
                Context("should successfully deploy all the resources", func() {
-                       test := func(withConfig bool, withWorkerConfig bool, runtimeVersionGreaterEquals126 bool) {
+                       test := func(withConfig bool, withWorkerConfig bool, runtimeVersionGreaterEquals126, withPriorityExpander bool) {
                                var config *gardencorev1beta1.ClusterAutoscaler
                                var shootWorkerConfig []gardencorev1beta1.Worker
                                if withConfig {
-                                       config = configFull
+                                       // Copy `configFull` so that the test doesn't overwrite it.
+                                       config = configFull.DeepCopy()
+                                       if withPriorityExpander {
+                                               config.Expander = ptr.To(gardencorev1beta1.ClusterAutoscalerExpanderPriority + "," + configExpander)
+                                       }
                                }
 
                                if withWorkerConfig {
@@ -709,11 +712,11 @@ var _ = Describe("ClusterAutoscaler", func() {
                                Expect(actualServiceMonitor).To(DeepEqual(serviceMonitor))
                        }
 
-                       It("w/o config", func() { test(false, false, false) })
-                       It("w/ config, kubernetes version < 1.26", func() { test(true, false, false) })
-                       It("w/ config, kubernetes version >= 1.26", func() { test(true, false, true) })
-                       It("w/ config, w/ workerConfig, kubernetes version >= 1.26", func() { test(true, true, true) })
-
+                       It("w/o config", func() { test(false, false, false, false) })
+                       It("w/ config, kubernetes version < 1.26", func() { test(true, false, false, false) })
+                       It("w/ config, kubernetes version >= 1.26", func() { test(true, false, true, false) })
+                       It("w/ config, w/ workerConfig, kubernetes version >= 1.26", func() { test(true, true, true, false) })
+                       It("w/ config, w/ workerConfig, kubernetes version >= 1.26, w/ 'priority' expander already configured", func() { test(true, true, true, true) })
                })
        })
```

Reply:
Alright, sounds good :)
Thank you for the insight!